### PR TITLE
Hangman game/refactor

### LINKED ISSUE #4: Refactor
Control of the game state should be left to `HangmanGame`. `GuessInput` shouldn't care if the guess is a good one or a bad one, and shouldn't need to know what the word is. It can still handle clearing the input etc, cos that is part of its function.
​
In this case, since the functionality actually exists in `HangmanGame`, but is (currently) passed down to `GuessInput`, we only need to move the logic of the `handleGuess` function.
​
BUT, in order to do that, you need to test the current functionality of `HangmanGame` as it relates to an input. So, each of the contexts you wrote for `GuessInput` will need to be matched for `HangmanGame`. Of course, the outcomes will be different, but the contexts are the same.
​
Once that is done, you can lift much of `handleGuess` out of `GuessInput` and in to the game. You should be able to do this without breaking any of `HangmanGame`'s tests.
​
Do the whole thing as one PR, but I'd like to see a commit that has just the new tests, then one (or many) with the shift, and new tests for `GuessInput`. (There should be only 2 or 3 at that point)
​
At the end `GuessInput` should take only `addGuess` (or something like that) as a param.
​
This refactor should include 
[extract method](https://refactoring.guru/extract-method) and 
[move method](https://refactoring.guru/move-method)

### PR DESCRIPTION
**Pull request by CoreyGinnivan**
- Moved core functionality from GuessInput into HangmanGame
- Created main HangmanGame tests, and added new coverage for added functionality
- Updated GuessInput tests to cover what's controlled inside the component

Resolves #4

### CODE REVIEW

File: src/components/GuessInput.jsx [hunk 1]
```diff
@@ -1,26 +1,8 @@
-import { useState } from "react";
-
-export function GuessInput({ setGuesses, guesses, setFails, fails, word }) {
-  const [inputLetter, setInputLetter] = useState("");
-
+export function GuessInput({ addGuess, inputLetter, setInputLetter }) {
```
Comment (HookyQR) [inline src/components/GuessInput.jsx:1]: This should still handle `inputLetter` internally. That's its one job. It should call `addGuess` with `inputLetter` (currently called with nothing)

Comment (HookyQR) [reply src/components/GuessInput.jsx:1]: A nice wrapper inside this component would help:

```js
export function GuessInput({ addGuess }) {
  const onClick = () => addGuess(inputLetter)
  // ...
          if (event.key === "Enter") { onClick() }
  // ...
        <button {...{className: "guessinput__submitbutton", onClick, disabled}} >
```

Which brings me to `isDisabled`. Is there any reason this isn't just `disabled`? Doesn't make it any less readable, and makes it easier to use. Not like you've got an `isntDisabled` lying around that you need to contrast with.

File: src/components/__tests__/hangman_game.test.jsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+          component.find(".guessinput__submitbutton").simulate("click");
+        });
+
+        expect(alert).toHaveBeenCalledWith('You already guessed "L"!');
+      });
+    });
+
+    describe("when the guess input is empty", () => {
+      it("shows an alert if the guess is attempted to be submitted with the keyboard", () => {
+        const alert = (global.alert = jest.fn());
```
Comment (HookyQR) [inline src/components/__tests__/hangman_game.test.jsx:105]: You shouldn't have to set `const alert` here. `alert` === `global.alert` unless defined somewhere in the scope.

File: src/components/__tests__/hangman_game.test.jsx [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    />
+  );
+
+  describe("to play the game", () => {
+    it("initially needs to render the guessing component", () => {
+      expect(component.find("GuessInput")).toExist();
+      expect(component.find("GameOverScreen")).not.toExist();
+    });
+
+    it("needs to have a masked word and render same amount of blank spaces", () => {
```
Comment (HookyQR) [inline src/components/__tests__/hangman_game.test.jsx:32]:
```suggestion
    it("has a masked word and render same amount of blank spaces", () => {
```
Be present. Be one with your tests.

**Review by HookyQR — CHANGES_REQUESTED** [review]
Looking better. Now give `GuessInput` back it's letter to handle. 😉

File: src/components/GuessInput.jsx (hunk 1 above)
Comment (CoreyGinnivan) [reply src/components/GuessInput.jsx:1]: @HookyQR A lil confused with this one 😬 so I'm keeping the function of `addGuess` still in the main game, but I'm putting the state management back inside the GuessInput component? `const [inputLetter, setInputLetter] = useState("");`

Comment (CoreyGinnivan) [reply src/components/GuessInput.jsx:1]: Think I just had to rubber duck it 😅

Comment (HookyQR) [reply src/components/GuessInput.jsx:1]: addGuess belongs to the game. All things pertaining to the current letter belong to `GuessInput`. (Doesn't become part of the game's data until it is submitted)

**Comment by HookyQR** [comment]
You know what. The one message I didn't get was the review request. 🤦

File: src/components/GuessInput.jsx [hunk 4]
```diff
@@ -1,27 +1,11 @@
-import { useState } from "react";
-
-export function GuessInput({ setGuesses, guesses, setFails, fails, word }) {
-  const [inputLetter, setInputLetter] = useState("");
+export function GuessInput({ addGuess, inputLetter, setInputLetter }) {
```
Comment (HookyQR) [inline src/components/GuessInput.jsx:1]: The signature of this component should be:
```
GuessInput({ addGuess })
```
It should contain the `inputLetter` management in its entirety. The game does not need, and should not want anything to do with the 'input'. Only the 'guess'.

**Review by HookyQR — CHANGES_REQUESTED** [review]
Getting closer.

File: src/components/GuessInput.jsx [hunk 5]
```diff
@@ -1,6 +1,9 @@
-export function GuessInput({ addGuess, inputLetter, setInputLetter }) {
-  const onClick = () => addGuess(inputLetter);
+import { useState } from "react";
 
+export function GuessInput({ addGuess }) {
+  const [inputLetter, setInputLetter] = useState("");
+
+  const onClick = () => addGuess(inputLetter, setInputLetter);
```
Comment (HookyQR) [inline src/components/GuessInput.jsx:6]:
```js
  const onClick = () => addGuess(inputLetter);
```

The `setLetterGuess` at the end is not required. There's no need to clear that input, as any other input will clear it for you. If you really think that's required, you can just:

```js
  const onClick = () => {
    addGuess(inputLetter);
    setInputLetter("");
  }
```

I'll say it again... The game does not need to know how the input works. It only has to handle the submission of a guess, just the letter. Not handle what happens to the input box after the fact, no handle clearing the stored input. It does not, and should not care where the letter comes from or how, `addGuess` takes a letter and updates the _game state_ with it, that is all.

File: src/components/HangmanGame.jsx [hunk 6]
```diff
@@ -26,18 +25,18 @@ export default function HangmanGame({
         .join(" ")
     : "";
 
-  function addGuess() {
-    if (!inputLetter) {
+  function addGuess(letterGuess, setLetterGuess) {
```
Comment (HookyQR) [inline src/components/HangmanGame.jsx:28]: You're being a little too verbose. A `letter` and a `guess` are the same thing coming in to this function, no need to call it both things. Pick one. 😉

File: src/components/HangmanGame.jsx [hunk 7]
```diff
@@ -47,7 +46,7 @@ export default function HangmanGame({
   return (
     <div className="hangman__wrapper">
       {!isGameOver ? (
-        <GuessInput {...{ addGuess, inputLetter, setInputLetter }} />
+        <GuessInput {...{ addGuess }} />
```
Comment (HookyQR) [inline src/components/HangmanGame.jsx:49]: Better, but you're just passing the knowledge back up now, as apposed to passing it down like you were before.

The game does not need to know about how a letter is created or cleared. It just needs the letter once it is ready.

**Review by HookyQR — CHANGES_REQUESTED** [review]
Too much knowledge being passed around.

**Review by HookyQR — APPROVED** [review]
Yep, let's do it. Don't forget to get rid of the `backspace`/`delete` handling on the cleanup PR. (AND TEST FIRST!!)

**Pull request merged**